File: src/lr.pl
#!/usr/bin/perl -w
#===============================================================================
# Wrapper for the Binaries
# Running solely on Loglikelihood Ratio
#===============================================================================
use strict;
use Data::Dumper;
use Carp;
use File::Basename;
use Cwd;

my $codepath = "./src";

my $file      = $ARGV[0] || "small-len10-50.txt";
my $mm        =  1;

# This parameter only affect when mm = 2 
# This values means that RECOUNT will only consider those
# neighboring tags with error probability greater than 0.00262689
my $minBaseEr = 0.00262689; 


if (scalar @ARGV != 1) {
     die "perl recount.pl <preprocessed_input>\n";
}

my $base      = basename( $file, ".txt" );
my $dir       = dirname($file)."/";
my $nb_file   = $dir . $base . ".nb";
my $prop_file = $dir . $base . ".prop";
my $nbq_file  = $dir . $base . ".nbq";
my $cwd       = getcwd;

#print STDERR "$dir $cwd\n";

my $fn    = $codepath ."/FindNeighboursWithQual";
my $gp    = $codepath ."/GenerateProportion";
my $em    = $codepath ."/EstimateTrueCount_llratio";


#print STDERR "FindNeighboursWithQual";
system("$fn $file $mm $minBaseEr");

#print STDERR "\nGenerateProportion";
system("$gp $file > $prop_file");

#print STDERR "\nComputing Predicted Count and LLRatio...\n";
system("$em $file");
print "\n";

unlink($nb_file);
unlink($prop_file);
unlink($nbq_file);

File: src/lr.py
#!/usr/bin/env python3
"""
Optimized Python port of lr.pl
Wrapper for log-likelihood ratio analysis

Copyright 2009-2025, Edward Wijaya
Ported and optimized to Python 2025
"""

import sys
import subprocess
import argparse
from pathlib import Path
from typing import List


def cleanup_files(*files: Path) -> None:
    """Remove multiple files silently."""
    for filepath in files:
        filepath.unlink(missing_ok=True)


def main():
    parser = argparse.ArgumentParser(
        description='Log-Likelihood Ratio Analysis - Optimized Python version'
    )
    parser.add_argument(
        'input_file',
        type=Path,
        nargs='?',
        default=Path("small-len10-50.txt"),
        help='Preprocessed input file (default: small-len10-50.txt)'
    )

    args = parser.parse_args()

    # Validate input
    if not args.input_file.exists():
        print(f"Error: Input file '{args.input_file}' not found", file=sys.stderr)
        sys.exit(1)

    # Setup paths
    codepath = Path("./src")
    input_file = args.input_file
    mm = 1
    min_base_error = 0.00262689

    base = input_file.stem
    directory = input_file.parent if input_file.parent.name else Path(".")

    # File paths
    nb_file = directory / f"{base}.nb"
    prop_file = directory / f"{base}.prop"
    nbq_file = directory / f"{base}.nbq"

    # Binary paths
    fn = codepath / "FindNeighboursWithQual"
    gp = codepath / "GenerateProportion"
    em = codepath / "EstimateTrueCount_llratio"

    try:
        # Step 1: Find neighbours with quality
        subprocess.run([str(fn), str(input_file), str(mm), str(min_base_error)], check=True)

        # Step 2: Generate proportion
        with prop_file.open('w') as outfile:
            subprocess.run([str(gp), str(input_file)], stdout=outfile, check=True)

        # Step 3: Compute predicted count and LLRatio
        subprocess.run([str(em), str(input_file)], check=True)
        print()  # Add newline

    finally:
        # Cleanup temporary files
        cleanup_files(nb_file, prop_file, nbq_file)


if __name__ == "__main__":
    main()
